File: infrastructure/mcp/server/src/universe.ts
// SPDX-License-Identifier: MIT
//
// Universe helpers — tiny fs-layer that maps Eustress's file-system-first
// conventions onto ergonomic tool inputs. Every function takes a
// pre-validated absolute path; the `resolveInUniverse` gatekeeper in
// index.ts is responsible for the path-safety check.

import * as fs from 'node:fs/promises';
import * as fsSync from 'node:fs';
import * as path from 'node:path';

// Cap result sizes so we never hand back megabytes to the client. MCP
// tools that blow past their natural bounds get truncated by the client
// silently — bounding up front gives us a clean "truncated" signal we
// can include in the response.
export const MAX_LIST_ITEMS = 500;
export const MAX_SEARCH_MATCHES = 200;
export const MAX_FILE_BYTES = 256 * 1024;

export type SpaceInfo = { name: string; path: string };
export type ScriptInfo = {
  name: string;
  folder: string;
  space: string;
  class: string;         // "Script" | "SoulScript" | "LocalScript" | "ModuleScript"
  sourcePath: string;    // absolute
  summaryPath: string | null;
};
export type EntityMatch = {
  name: string;
  class: string;
  space: string;
  path: string;          // absolute to _instance.toml
};

/**
 * Enumerate every Space under `{universe}/Spaces/`. Returns an empty list
 * if the directory doesn't exist (valid for a freshly-created universe).
 */
export async function listSpaces(universe: string): Promise<SpaceInfo[]> {
  const spacesDir = path.join(universe, 'Spaces');
  let entries: string[] = [];
  try {
    entries = await fs.readdir(spacesDir);
  } catch { return []; }
  const out: SpaceInfo[] = [];
  for (const name of entries) {
    const abs = path.join(spacesDir, name);
    try {
      const stat = await fs.stat(abs);
      if (stat.isDirectory()) {
        out.push({ name, path: abs });
      }
    } catch { /* broken symlink or race — skip */ }
  }
  return out.sort((a, b) => a.name.localeCompare(b.name));
}

/**
 * Walk one Space (or all Spaces) and return every folder-based Rune
 * script. Recognises a script by the presence of an `_instance.toml`
 * whose `class_name` sits in the script family AND a source file in the
 * same folder (canonical `<folder>.rune` or legacy `Source.rune`).
 */
export async function listScripts(
  universe: string,
  spaceName?: string,
): Promise<ScriptInfo[]> {
  const spaces = spaceName
    ? (await listSpaces(universe)).filter(s => s.name === spaceName)
    : await listSpaces(universe);

  const out: ScriptInfo[] = [];
  for (const space of spaces) {
    await walkForScripts(space.path, space.name, out);
    if (out.length >= MAX_LIST_ITEMS) break;
  }
  return out.slice(0, MAX_LIST_ITEMS);
}

async function walkForScripts(
  dir: string,
  spaceName: string,
  out: ScriptInfo[],
): Promise<void> {
  if (out.length >= MAX_LIST_ITEMS) return;
  let entries: fsSync.Dirent[] = [];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch { return; }

  // Is this directory itself a script folder?
  const instanceToml = entries.find(e => e.isFile() && e.name === '_instance.toml');
  if (instanceToml) {
    const info = await maybeReadAsScript(dir, spaceName);
    if (info) out.push(info);
  }

  // Recurse into subdirectories. We only skip `.eustress/` and hidden
  // folders — everything else can host scripts.
  for (const e of entries) {
    if (!e.isDirectory()) continue;
    if (e.name === '.eustress' || e.name.startsWith('.')) continue;
    if (out.length >= MAX_LIST_ITEMS) return;
    await walkForScripts(path.join(dir, e.name), spaceName, out);
  }
}

async function maybeReadAsScript(
  folder: string,
  space: string,
): Promise<ScriptInfo | null> {
  const tomlPath = path.join(folder, '_instance.toml');
  let toml: string;
  try {
    toml = await fs.readFile(tomlPath, 'utf8');
  } catch { return null; }

  const klass = extractToml(toml, 'class_name');
  const scriptClasses = ['Script', 'SoulScript', 'LocalScript', 'ModuleScript'];
  if (!klass || !scriptClasses.includes(klass)) return null;

  const name = path.basename(folder);
  const sourcePath = await findScriptSource(folder, name);
  if (!sourcePath) return null;

  const summaryPath = await findScriptSummary(folder, name);

  return {
    name, folder, space,
    class: klass,
    sourcePath,
    summaryPath,
  };
}

/** Canonical `<folder>/<folder>.rune` preferred, legacy fallbacks otherwise. */
async function findScriptSource(folder: string, name: string): Promise<string | null> {
  const canonical = path.join(folder, `${name}.rune`);
  if (await exists(canonical)) return canonical;

  let entries: fsSync.Dirent[] = [];
  try {
    entries = await fs.readdir(folder, { withFileTypes: true });
  } catch { return null; }
  for (const e of entries) {
    if (!e.isFile()) continue;
    if (e.name.endsWith('.rune') || e.name.endsWith('.luau')
        || e.name.endsWith('.soul') || e.name.endsWith('.lua')) {
      return path.join(folder, e.name);
    }
  }
  return null;
}

async function findScriptSummary(folder: string, name: string): Promise<string | null> {
  const canonical = path.join(folder, `${name}.md`);
  if (await exists(canonical)) return canonical;
  const legacy = path.join(folder, 'Summary.md');
  if (await exists(legacy)) return legacy;
  return null;
}

async function exists(p: string): Promise<boolean> {
  try { await fs.access(p); return true; } catch { return false; }
}

/**
 * Find entities by name across a Space (or all Spaces). Matches the `name`
 * field in any `_instance.toml`. Case-insensitive substring.
 */
export async function findEntity(
  universe: string,
  query: string,
  spaceName?: string,
): Promise<EntityMatch[]> {
  const needle = query.toLowerCase();
  const spaces = spaceName
    ? (await listSpaces(universe)).filter(s => s.name === spaceName)
    : await listSpaces(universe);

  const out: EntityMatch[] = [];
  for (const space of spaces) {
    await walkForEntities(space.path, space.name, needle, out);
    if (out.length >= MAX_LIST_ITEMS) break;
  }
  return out.slice(0, MAX_LIST_ITEMS);
}

async function walkForEntities(
  dir: string,
  spaceName: string,
  needle: string,
  out: EntityMatch[],
): Promise<void> {
  if (out.length >= MAX_LIST_ITEMS) return;
  let entries: fsSync.Dirent[] = [];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch { return; }

  for (const e of entries) {
    if (e.isDirectory()) {
      if (e.name === '.eustress' || e.name.startsWith('.')) continue;
      await walkForEntities(path.join(dir, e.name), spaceName, needle, out);
      continue;
    }
    if (!e.isFile()) continue;
    // We scan `_instance.toml` for folder-based entities, and any
    // `*.part.toml`, `*.glb.toml`, `*.model.toml` etc. as flat-file entities.
    const isInstance = e.name === '_instance.toml';
    const isFlatEntity = e.name.endsWith('.part.toml')
      || e.name.endsWith('.glb.toml')
      || e.name.endsWith('.model.toml')
      || e.name.endsWith('.textlabel.toml');
    if (!isInstance && !isFlatEntity) continue;

    const full = path.join(dir, e.name);
    let toml: string;
    try { toml = await fs.readFile(full, 'utf8'); } catch { continue; }

    const nameField = extractToml(toml, 'name');
    const klass = extractToml(toml, 'class_name') ?? 'Instance';
    // For folder-based entities the identity is the folder name; `name`
    // in the TOML is sometimes absent.
    const identity = nameField ?? (isInstance ? path.basename(dir) : e.name.split('.')[0]);
    if (!identity.toLowerCase().includes(needle)) continue;

    out.push({
      name: identity,
      class: klass,
      space: spaceName,
      path: full,
    });
    if (out.length >= MAX_LIST_ITEMS) return;
  }
}

export type SearchMatch = {
  path: string;
  line: number;          // 1-based
  preview: string;       // the matching line, trimmed
};

/**
 * Plain-text search across `.rune` and `.toml` files under the Universe.
 * Case-insensitive. Caps at MAX_SEARCH_MATCHES matches.
 */
export async function searchUniverse(
  universe: string,
  query: string,
  spaceName?: string,
): Promise<SearchMatch[]> {
  const roots = spaceName
    ? (await listSpaces(universe)).filter(s => s.name === spaceName).map(s => s.path)
    : [path.join(universe, 'Spaces')];

  const needle = query.toLowerCase();
  const out: SearchMatch[] = [];
  for (const root of roots) {
    await walkForSearch(root, needle, out);
    if (out.length >= MAX_SEARCH_MATCHES) break;
  }
  return out.slice(0, MAX_SEARCH_MATCHES);
}

async function walkForSearch(
  dir: string,
  needle: string,
  out: SearchMatch[],
): Promise<void> {
  if (out.length >= MAX_SEARCH_MATCHES) return;
  let entries: fsSync.Dirent[] = [];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch { return; }

  for (const e of entries) {
    const abs = path.join(dir, e.name);
    if (e.isDirectory()) {
      if (e.name === '.eustress' || e.name.startsWith('.')) continue;
      await walkForSearch(abs, needle, out);
      if (out.length >= MAX_SEARCH_MATCHES) return;
      continue;
    }
    if (!e.isFile()) continue;
    if (!e.name.endsWith('.rune') && !e.name.endsWith('.toml')
        && !e.name.endsWith('.md')) continue;

    let contents: string;
    try { contents = await fs.readFile(abs, 'utf8'); } catch { continue; }
    const lines = contents.split('\n');
    for (let i = 0; i < lines.length; i++) {
      if (lines[i].toLowerCase().includes(needle)) {
        out.push({ path: abs, line: i + 1, preview: lines[i].trim().slice(0, 200) });
        if (out.length >= MAX_SEARCH_MATCHES) return;
      }
    }
  }
}

/**
 * Extract a scalar string field from a TOML file. Not a real parser —
 * handles the single-value lines Eustress writes (`name = "foo"`,
 * `class_name = "Script"`). Returns `null` if not found.
 */
export function extractToml(toml: string, field: string): string | null {
  for (const line of toml.split('\n')) {
    const trimmed = line.trim();
    // match "field = "value"" — quotes required for string values
    const prefix = `${field} = "`;
    if (trimmed.startsWith(prefix)) {
      const rest = trimmed.slice(prefix.length);
      const end = rest.indexOf('"');
      if (end >= 0) return rest.slice(0, end);
    }
  }
  return null;
}

/**
 * Read a file with a byte cap so we don't blow up the MCP client with
 * 100 KB of unwanted content. Returns `{text, truncated}` — caller
 * includes the truncation flag in their response.
 */
export async function readCapped(
  abs: string,
): Promise<{ text: string; truncated: boolean }> {
  const stat = await fs.stat(abs);
  const truncated = stat.size > MAX_FILE_BYTES;
  const buf = await fs.readFile(abs);
  const slice = truncated ? buf.subarray(0, MAX_FILE_BYTES) : buf;
  return { text: slice.toString('utf8'), truncated };
}

/**
 * Path-safety gatekeeper. Resolves `input` against `universe`, rejects
 * anything that escapes the root. Returns the resolved absolute path on
 * success; throws a string on violation so the caller can surface a
 * meaningful MCP error.
 */
export function resolveInUniverse(universe: string, input: string): string {
  const universeAbs = path.resolve(universe);
  const resolved = path.isAbsolute(input)
    ? path.resolve(input)
    : path.resolve(universeAbs, input);
  const rel = path.relative(universeAbs, resolved);
  if (rel.startsWith('..') || path.isAbsolute(rel)) {
    throw `path '${input}' escapes the Universe root`;
  }
  return resolved;
}

// ─── Dynamic Universe discovery ──────────────────────────────────────

/**
 * Walk up from `startPath` looking for the enclosing Universe — any
 * directory that has both a `Spaces/` subdirectory and (optionally) a
 * `.eustress/` marker. Stops at filesystem root.
 *
 * The "has Spaces/" check alone is enough to call a folder a Universe
 * at the fs layer; `.eustress/` is a nice-to-have that signals the
 * engine has actually initialised there.
 *
 * Returns `null` if no Universe is found anywhere between `startPath`
 * and `/` — callers typically fall back to an explicit config or bail.
 */
export function findUniverseRoot(startPath: string): string | null {
  let cur = path.resolve(startPath);
  const seen = new Set<string>();
  // Bounded walk so a pathological cycle (Windows junction loops) can't
  // hang the tool call.
  for (let i = 0; i < 64; i++) {
    if (seen.has(cur)) return null;
    seen.add(cur);
    const spaces = path.join(cur, 'Spaces');
    if (fsSync.existsSync(spaces) && fsSync.statSync(spaces).isDirectory()) {
      return cur;
    }
    const parent = path.dirname(cur);
    if (parent === cur) return null;  // hit filesystem root
    cur = parent;
  }
  return null;
}

/**
 * Shallow scan of `roots` for directories that look like Universes.
 * Used by `eustress_list_universes`. Each root is checked one level
 * deep — `roots[i]/<child>/Spaces/` — which matches how users typically
 * organise multiple projects (e.g. `~/Eustress/MyGame/Spaces/`,
 * `~/Eustress/SideProject/Spaces/`).
 *
 * Also returns any root that is itself a Universe.
 */
export async function discoverUniverses(roots: string[]): Promise<string[]> {
  const out = new Set<string>();

  for (const raw of roots) {
    const root = path.resolve(raw);
    if (fsSync.existsSync(path.join(root, 'Spaces'))) {
      out.add(root);
    }
    let entries: fsSync.Dirent[] = [];
    try {
      entries = await fs.readdir(root, { withFileTypes: true });
    } catch { continue; }
    for (const e of entries) {
      if (!e.isDirectory() || e.name.startsWith('.')) continue;
      const child = path.join(root, e.name);
      if (fsSync.existsSync(path.join(child, 'Spaces'))) {
        out.add(child);
      }
    }
  }
  return Array.from(out).sort();
}

File: eustress/crates/mcp-server/src/universe.rs
// Universe helpers — tiny fs layer that maps Eustress's file-system-first
// conventions onto ergonomic tool inputs. Functions take pre-validated
// absolute paths; `resolve_in_universe` is the gatekeeper for path safety.

use serde::Serialize;
use std::collections::HashSet;
use std::path::{Component, Path, PathBuf};

// Cap result sizes so we never hand back megabytes to the client. MCP clients
// that blow past their natural bounds get truncated silently on their side;
// bounding up front gives us a clean "truncated" signal we can surface.
pub const MAX_LIST_ITEMS: usize = 500;
pub const MAX_SEARCH_MATCHES: usize = 200;
pub const MAX_FILE_BYTES: usize = 256 * 1024;

#[derive(Serialize, Debug, Clone)]
pub struct SpaceInfo {
    pub name: String,
    pub path: String,
}

#[derive(Serialize, Debug, Clone)]
pub struct ScriptInfo {
    pub name: String,
    pub folder: String,
    pub space: String,
    pub class: String,
    pub source_path: String,
    pub summary_path: Option<String>,
}

#[derive(Serialize, Debug, Clone)]
pub struct EntityMatch {
    pub name: String,
    pub class: String,
    pub space: String,
    pub path: String,
}

#[derive(Serialize, Debug, Clone)]
pub struct SearchMatch {
    pub path: String,
    pub line: usize,
    pub preview: String,
}

pub fn list_spaces(universe: &Path) -> Vec<SpaceInfo> {
    let spaces_dir = universe.join("Spaces");
    let read = match std::fs::read_dir(&spaces_dir) {
        Ok(r) => r,
        Err(_) => return Vec::new(),
    };
    let mut out = Vec::new();
    for entry in read.flatten() {
        if let Ok(ft) = entry.file_type() {
            if ft.is_dir() {
                let name = entry.file_name().to_string_lossy().to_string();
                let path = entry.path().to_string_lossy().to_string();
                out.push(SpaceInfo { name, path });
            }
        }
    }
    out.sort_by(|a, b| a.name.cmp(&b.name));
    out
}

pub fn list_scripts(universe: &Path, space_filter: Option<&str>) -> Vec<ScriptInfo> {
    let spaces: Vec<SpaceInfo> = list_spaces(universe)
        .into_iter()
        .filter(|s| space_filter.map_or(true, |f| f == s.name))
        .collect();

    let mut out = Vec::new();
    for space in &spaces {
        walk_for_scripts(Path::new(&space.path), &space.name, &mut out);
        if out.len() >= MAX_LIST_ITEMS {
            break;
        }
    }
    out.truncate(MAX_LIST_ITEMS);
    out
}

fn walk_for_scripts(dir: &Path, space_name: &str, out: &mut Vec<ScriptInfo>) {
    if out.len() >= MAX_LIST_ITEMS {
        return;
    }
    let read = match std::fs::read_dir(dir) {
        Ok(r) => r,
        Err(_) => return,
    };
    let entries: Vec<_> = read.flatten().collect();

    // Is this directory itself a script folder?
    let has_instance = entries
        .iter()
        .any(|e| e.file_name() == std::ffi::OsStr::new("_instance.toml"));
    if has_instance {
        if let Some(info) = maybe_read_as_script(dir, space_name) {
            out.push(info);
        }
    }

    for e in &entries {
        let ft = match e.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };
        if !ft.is_dir() {
            continue;
        }
        let name = e.file_name().to_string_lossy().to_string();
        if name == ".eustress" || name.starts_with('.') {
            continue;
        }
        if out.len() >= MAX_LIST_ITEMS {
            return;
        }
        walk_for_scripts(&e.path(), space_name, out);
    }
}

fn maybe_read_as_script(folder: &Path, space: &str) -> Option<ScriptInfo> {
    let toml_path = folder.join("_instance.toml");
    let toml_text = std::fs::read_to_string(&toml_path).ok()?;
    let klass = extract_toml(&toml_text, "class_name")?;
    const SCRIPT_CLASSES: &[&str] = &["Script", "SoulScript", "LocalScript", "ModuleScript"];
    if !SCRIPT_CLASSES.contains(&klass.as_str()) {
        return None;
    }

    let name = folder.file_name()?.to_string_lossy().to_string();
    let source_path = find_script_source(folder, &name)?;
    let summary_path = find_script_summary(folder, &name);

    Some(ScriptInfo {
        name,
        folder: folder.to_string_lossy().to_string(),
        space: space.to_string(),
        class: klass,
        source_path: source_path.to_string_lossy().to_string(),
        summary_path: summary_path.map(|p| p.to_string_lossy().to_string()),
    })
}

fn find_script_source(folder: &Path, name: &str) -> Option<PathBuf> {
    let canonical = folder.join(format!("{name}.rune"));
    if canonical.exists() {
        return Some(canonical);
    }
    for entry in std::fs::read_dir(folder).ok()?.flatten() {
        let file_name = entry.file_name();
        let name_str = file_name.to_string_lossy();
        if !entry.file_type().ok()?.is_file() {
            continue;
        }
        if name_str.ends_with(".rune")
            || name_str.ends_with(".luau")
            || name_str.ends_with(".soul")
            || name_str.ends_with(".lua")
        {
            return Some(entry.path());
        }
    }
    None
}

fn find_script_summary(folder: &Path, name: &str) -> Option<PathBuf> {
    let canonical = folder.join(format!("{name}.md"));
    if canonical.exists() {
        return Some(canonical);
    }
    let legacy = folder.join("Summary.md");
    if legacy.exists() {
        return Some(legacy);
    }
    None
}

pub fn find_entity(universe: &Path, query: &str, space_filter: Option<&str>) -> Vec<EntityMatch> {
    let needle = query.to_lowercase();
    let spaces: Vec<SpaceInfo> = list_spaces(universe)
        .into_iter()
        .filter(|s| space_filter.map_or(true, |f| f == s.name))
        .collect();

    let mut out = Vec::new();
    for space in &spaces {
        walk_for_entities(Path::new(&space.path), &space.name, &needle, &mut out);
        if out.len() >= MAX_LIST_ITEMS {
            break;
        }
    }
    out.truncate(MAX_LIST_ITEMS);
    out
}

fn walk_for_entities(dir: &Path, space_name: &str, needle: &str, out: &mut Vec<EntityMatch>) {
    if out.len() >= MAX_LIST_ITEMS {
        return;
    }
    let entries = match std::fs::read_dir(dir) {
        Ok(r) => r,
        Err(_) => return,
    };

    for e in entries.flatten() {
        let ft = match e.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };
        let name = e.file_name().to_string_lossy().to_string();
        if ft.is_dir() {
            if name == ".eustress" || name.starts_with('.') {
                continue;
            }
            walk_for_entities(&e.path(), space_name, needle, out);
            if out.len() >= MAX_LIST_ITEMS {
                return;
            }
            continue;
        }
        if !ft.is_file() {
            continue;
        }
        let is_instance = name == "_instance.toml";
        let is_flat = name.ends_with(".part.toml")
            || name.ends_with(".glb.toml")
            || name.ends_with(".model.toml")
            || name.ends_with(".textlabel.toml");
        if !is_instance && !is_flat {
            continue;
        }
        let full = e.path();
        let toml_text = match std::fs::read_to_string(&full) {
            Ok(t) => t,
            Err(_) => continue,
        };
        let name_field = extract_toml(&toml_text, "name");
        let klass = extract_toml(&toml_text, "class_name").unwrap_or_else(|| "Instance".to_string());
        let identity = match name_field {
            Some(n) => n,
            None if is_instance => dir.file_name().map(|s| s.to_string_lossy().to_string()).unwrap_or_default(),
            None => name.split('.').next().unwrap_or(&name).to_string(),
        };
        if !identity.to_lowercase().contains(needle) {
            continue;
        }
        out.push(EntityMatch {
            name: identity,
            class: klass,
            space: space_name.to_string(),
            path: full.to_string_lossy().to_string(),
        });
        if out.len() >= MAX_LIST_ITEMS {
            return;
        }
    }
}

pub fn search_universe(universe: &Path, query: &str, space_filter: Option<&str>) -> Vec<SearchMatch> {
    let needle = query.to_lowercase();
    let roots: Vec<PathBuf> = if let Some(space) = space_filter {
        list_spaces(universe)
            .into_iter()
            .filter(|s| s.name == space)
            .map(|s| PathBuf::from(s.path))
            .collect()
    } else {
        vec![universe.join("Spaces")]
    };

    let mut out = Vec::new();
    for root in &roots {
        walk_for_search(root, &needle, &mut out);
        if out.len() >= MAX_SEARCH_MATCHES {
            break;
        }
    }
    out.truncate(MAX_SEARCH_MATCHES);
    out
}

fn walk_for_search(dir: &Path, needle: &str, out: &mut Vec<SearchMatch>) {
    if out.len() >= MAX_SEARCH_MATCHES {
        return;
    }
    let entries = match std::fs::read_dir(dir) {
        Ok(r) => r,
        Err(_) => return,
    };
    for e in entries.flatten() {
        let ft = match e.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };
        let name = e.file_name().to_string_lossy().to_string();
        let abs = e.path();
        if ft.is_dir() {
            if name == ".eustress" || name.starts_with('.') {
                continue;
            }
            walk_for_search(&abs, needle, out);
            if out.len() >= MAX_SEARCH_MATCHES {
                return;
            }
            continue;
        }
        if !ft.is_file() {
            continue;
        }
        if !(name.ends_with(".rune") || name.ends_with(".toml") || name.ends_with(".md")) {
            continue;
        }
        let text = match std::fs::read_to_string(&abs) {
            Ok(t) => t,
            Err(_) => continue,
        };
        for (i, line) in text.lines().enumerate() {
            if line.to_lowercase().contains(needle) {
                let preview = line.trim();
                let capped: String = preview.chars().take(200).collect();
                out.push(SearchMatch {
                    path: abs.to_string_lossy().to_string(),
                    line: i + 1,
                    preview: capped,
                });
                if out.len() >= MAX_SEARCH_MATCHES {
                    return;
                }
            }
        }
    }
}

/// Extract a scalar string field from a TOML file. Uses the real `toml`
/// parser so nested tables and mixed content parse correctly (improvement
/// over the TS version's per-line substring hack).
pub fn extract_toml(text: &str, field: &str) -> Option<String> {
    let value: toml::Value = toml::from_str(text).ok()?;
    extract_toml_recursive(&value, field)
}

fn extract_toml_recursive(v: &toml::Value, field: &str) -> Option<String> {
    match v {
        toml::Value::Table(t) => {
            if let Some(found) = t.get(field).and_then(|x| x.as_str()) {
                return Some(found.to_string());
            }
            for (_, nested) in t.iter() {
                if let Some(hit) = extract_toml_recursive(nested, field) {
                    return Some(hit);
                }
            }
            None
        }
        _ => None,
    }
}

pub struct CappedRead {
    pub text: String,
    pub truncated: bool,
}

pub fn read_capped(abs: &Path) -> std::io::Result<CappedRead> {
    let bytes = std::fs::read(abs)?;
    let truncated = bytes.len() > MAX_FILE_BYTES;
    let slice = if truncated { &bytes[..MAX_FILE_BYTES] } else { &bytes[..] };
    let text = String::from_utf8_lossy(slice).into_owned();
    Ok(CappedRead { text, truncated })
}

/// Path-safety gatekeeper. Resolves `input` against `universe`, rejects
/// anything that escapes the root. Returns the resolved absolute path on
/// success, `Err(msg)` on violation.
pub fn resolve_in_universe(universe: &Path, input: &str) -> Result<PathBuf, String> {
    let universe_abs = universe
        .canonicalize()
        .unwrap_or_else(|_| universe.to_path_buf());
    let input_path = Path::new(input);
    let resolved = if input_path.is_absolute() {
        input_path.to_path_buf()
    } else {
        universe_abs.join(input_path)
    };
    // Canonicalize if it exists; otherwise normalize components manually so we
    // can still reject `..` escapes for paths we're about to create.
    let normalized = match resolved.canonicalize() {
        Ok(p) => p,
        Err(_) => normalize_components(&resolved),
    };
    if !normalized.starts_with(&universe_abs) {
        return Err(format!("path '{input}' escapes the Universe root"));
    }
    Ok(normalized)
}

fn normalize_components(p: &Path) -> PathBuf {
    let mut out = PathBuf::new();
    for c in p.components() {
        match c {
            Component::ParentDir => {
                out.pop();
            }
            Component::CurDir => {}
            other => out.push(other.as_os_str()),
        }
    }
    out
}

/// Walk up from `start` looking for the enclosing Universe — any directory
/// with a `Spaces/` subdirectory. Bounded to 64 iterations so a pathological
/// cycle (Windows junction loops) can't hang us.
pub fn find_universe_root(start: &Path) -> Option<PathBuf> {
    let mut cur = start.canonicalize().unwrap_or_else(|_| start.to_path_buf());
    let mut seen = HashSet::new();
    for _ in 0..64 {
        if !seen.insert(cur.clone()) {
            return None;
        }
        let spaces = cur.join("Spaces");
        if spaces.is_dir() {
            return Some(cur);
        }
        let parent = cur.parent()?.to_path_buf();
        if parent == cur {
            return None;
        }
        cur = parent;
    }
    None
}

/// Shallow scan of `roots` for Universes. Each root is itself checked, and
/// one level down is walked (matches how users organise: `~/Eustress/MyGame/`,
/// `~/Eustress/SideProject/`).
pub fn discover_universes(roots: &[PathBuf]) -> Vec<PathBuf> {
    let mut out = HashSet::new();
    for raw in roots {
        let root = raw.canonicalize().unwrap_or_else(|_| raw.clone());
        if root.join("Spaces").is_dir() {
            out.insert(root.clone());
        }
        let read = match std::fs::read_dir(&root) {
            Ok(r) => r,
            Err(_) => continue,
        };
        for e in read.flatten() {
            let ft = match e.file_type() {
                Ok(ft) => ft,
                Err(_) => continue,
            };
            if !ft.is_dir() {
                continue;
            }
            let name = e.file_name().to_string_lossy().to_string();
            if name.starts_with('.') {
                continue;
            }
            let child = root.join(&name);
            if child.join("Spaces").is_dir() {
                out.insert(child);
            }
        }
    }
    let mut v: Vec<PathBuf> = out.into_iter().collect();
    v.sort();
    v
}

/// Parse `EUSTRESS_UNIVERSES_PATH` (OS path-separator delimited).
pub fn parse_search_roots(env_val: Option<&str>) -> Vec<PathBuf> {
    if let Some(raw) = env_val {
        return std::env::split_paths(raw).collect();
    }
    let mut out = Vec::new();
    if let Some(home) = dirs::home_dir() {
        out.push(home.join("Eustress"));
        out.push(home.join("Documents").join("Eustress"));
        out.push(home);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extract_toml_top_level() {
        let src = "[metadata]\nclass_name = \"Script\"\nother = 1";
        assert_eq!(extract_toml(src, "class_name").as_deref(), Some("Script"));
    }

    #[test]
    fn extract_toml_missing() {
        let src = "[metadata]\nname = \"X\"";
        assert!(extract_toml(src, "class_name").is_none());
    }

    #[test]
    fn resolve_in_universe_rejects_escape() {
        let tmp = std::env::temp_dir();
        let u = tmp.join("eustress-test-universe");
        std::fs::create_dir_all(u.join("Spaces")).ok();
        let err = resolve_in_universe(&u, "../../etc/passwd");
        assert!(err.is_err());
    }
}
